File: src/lexer/lexer.cc
#include "monkey/lexer/lexer.hh"

#include <algorithm>
#include <map>
#include <optional>
#include <string>
#include <string_view>

#include "monkey/common.hh"
#include "monkey/lexer/token.hh"

BEGIN_MONKEY_NAMESPACE

Lexer::Lexer(const std::string_view _input) noexcept
  : input_{ std::move(_input) }
  , position_{ input_.begin() }
{
}

void
Lexer::emplace(const std::string_view _new_input) noexcept
{
    input_ = std::move(_new_input);
    position_ = input_.begin();
}

bool
Lexer::peek(const char _c) noexcept
{
    bool matched = *position_ == _c;

    if (matched) {
        position_ += 1;
    }

    return matched;
}

namespace detail {

constexpr bool
identifier(const char _c) noexcept
{
    return isalnum(_c) || _c == '_';
}

std::optional<char>
get_escape_sequence(char escaped) noexcept
{
    // clang-format off
    static const std::map<char, char> escape_sequence_map{
        {'a', '\a'},
        {'b', '\b'},
        {'f', '\f'},
        {'n', '\n'},
        {'r', '\r'},
        {'t', '\t'},
        {'v', '\v'},
        {'\\', '\\'},
        {'\'', '\''},
        {'\"', '\"'},
    };
    // clang-format on

    auto escape_sequence = escape_sequence_map.find(escaped);

    if (escape_sequence == escape_sequence_map.end()) {
        return std::nullopt;
    }

    return escape_sequence->second;
}

} // namespace detail

std::optional<Token>
Lexer::next() noexcept
{
    // Skip any whitespaces before the next token.
    position_ = std::find_if_not(position_, input_.end(), ::isspace);

    if (position_ >= input_.end()) {
        return std::nullopt; // We have reached the end of the file.
    }

    const std::string_view::const_iterator start_position = position_++;

    switch (*start_position) {
        /* Operators */
        case '=':
            return peek('=') ? Equal : Assign;
        case '!':
            return peek('=') ? NotEqual : Bang;
        case '+':
            return Plus;
        case '-':
            return Minus;
        case '*':
            return Asterisk;
        case '/':
            return Slash;
        case '<':
            return peek('=') ? LessEqual : LessThan;
        case '>':
            return peek('=') ? GreaterEqual : GreaterThan;
        /* Delimiters */
        case ',':
            return Comma;
        case ';':
            return Semicolon;
        case ':':
            return Colon;
        case '(':
            return LParenthesis;
        case ')':
            return RParenthesis;
        case '{':
            return LBrace;
        case '}':
            return RBrace;
        case '[':
            return LBracket;
        case ']':
            return RBracket;
        case '"':
            return get_string_token();
        default:
            break;
    }

    // Keyword or Identifier
    if (isalpha(*start_position) || *start_position == '_') {
        position_ = std::find_if_not(position_, input_.end(), detail::identifier);

        const std::string_view literal{ start_position, position_ };
        const auto keyword = keyword_map.find(literal);

        if (keyword != keyword_map.end()) {
            return Token{ keyword->second, std::move(literal) };
        }

        return Token{ TokenType::Identifier, std::move(literal) };
    }

    // Integer
    if (::isdigit(*start_position)) {
        position_ = std::find_if_not(position_, input_.end(), ::isdigit);
        std::string_view literal{ start_position, position_ };
        return Token{ TokenType::Integer, std::move(literal) };
    }

    // Unrecognized token
    return Token{ TokenType::Illegal, { start_position, position_ } };
}

// TODO: Throw exceptions for bad input?
std::optional<Token>
Lexer::get_string_token()
{
    std::string str{};
    char c = '\0';

    while (position_ < input_.end() && (c = *position_++) != '"') {
        if (c == '\\') {
            if (position_ == input_.end()) {
                return std::nullopt; // Unterminated string.
            }

            std::optional<char> escape = detail::get_escape_sequence(*position_++);

            if (!escape.has_value()) {
                return std::nullopt; // Invalid escape sequence.
            }

            str += escape.value();
        } else {
            str += c;
        }
    }

    if (c != '"') {
        return std::nullopt; // Unterminated string.
    }

    return Token{ TokenType::String, std::move(str) };
}

END_MONKEY_NAMESPACE

File: src/lexer/lexer.go
package lexer

import "monkey/token"

type Lexer struct {
	input        string
	position     int  // current position in input (points to current char)
	readPosition int  // current reading position in input (after current char)
	ch           byte // current char under examination
}

func New(input string) *Lexer {
	l := &Lexer{input: input}
	// Set the lexer to a working state before handing it off to the user.
	l.readChar()
	return l
}

// Return the next character and advance our position in the input string.
func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		// If we go off the end of the input, set `ch` to NULL.
		l.ch = 0
	} else {
		// Otherwise, read the next character.
		l.ch = l.input[l.readPosition]
	}

	// Advance both iterators by 1.
	l.position = l.readPosition
	l.readPosition += 1
}

func (l *Lexer) NextToken() token.Token {
	var tok token.Token

	switch l.ch {
	case '=':
		tok = newToken(token.ASSIGN, l.ch)
	case ';':
		tok = newToken(token.SEMICOLON, l.ch)
	case '(':
		tok = newToken(token.LPARENTHESIS, l.ch)
	case ')':
		tok = newToken(token.RPARENTHESIS, l.ch)
	case ',':
		tok = newToken(token.COMMA, l.ch)
	case '+':
		tok = newToken(token.PLUS, l.ch)
	case '{':
		tok = newToken(token.LBRACE, l.ch)
	case '}':
		tok = newToken(token.RBRACE, l.ch)
	case 0:
		tok.Literal = ""
		tok.Type = token.EOF
	}

	l.readChar()
	return tok
}

func newToken(tokenType token.TokenType, ch byte) token.Token {
	return token.Token{Type: tokenType, Literal: string(ch)}
}
